### ISSUE TITLE
C++20 Enhancements for SFML 4

### ISSUE THREAD

**Issue by ChrisThrasher** [SFML/SFML#3143]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I understand [SFML's scope](https://www.sfml-dev.org/contribute.php#general-considerations) and believe the feature fits it

### Describe your feature request here

SFML 4 will require C++20 or newer. This issue catalogs places where SFML could benefit from utilizing C++20 features. Some of these changes imply breaking the API while others are merely implementation improvements.

- Replace `sf::U8string` with `std::u8string`
- Use ranged algorithms (`std::ranges::{sort|copy|is_permutation|find|find_if|remove_if|is_sorted}`)
- Use `operator<=>` to define comparison operations
- Use `operator==() = default` to define (in)equality operations
- Use `<numbers>` for Pi constant
- Replace pointer/size parameter pairs with `std::span`

### Use Cases

N/A

### API Example

N/A

**Comment by ZXShady** [comment]
also changing Shader::setUniform to use string_view instead of const string& since C++20 allows heterogenous lookups.

**Comment by ZXShady** [comment]
sf::Vector2 can't be made an aggregate since it has another constructor that cannot be replaced

`SFML_SYSTEM_API Vector2(T r, Angle phi);`

**Comment by TerensTare** [comment]
Just brainstorming, I think it's better if this case can use a named constructor (ie. a static method that returns a vector), which makes `Vector2` an aggregate. The code would probably like something like this:
```cpp
template <typename T>
struct Vector2
{
    static constexpr Vector2 fromPolar(T r, Angle phi)
    {
        auto phi_r = phi.asRadians();
        auto cos = std::cosf(phi_r);
        auto sin = std::sinf(phi_r);
        return Vector2{.x = r * cos, .y = r * sin};
    }

    T x, y;
};
```

**Comment by ZXShady** [comment]
@TerensTare 


the uses of an aggregate Vector2 is not particularly helpful, as designated initializers really shine with multiple members not just 2 members that are known in order making it an aggregate really just saves us 10 lines of code or so.


also we also need to remove the conversion constructor to make it an aggregate and turn it into the explcicit conversion operator instead

```cpp

template<typename T>
struct Vector2 {
      constexpr explicit operator Vector2<U>() {
              return {U(x),U(y)};
      }
};

```

**Comment by ChrisThrasher** [comment]
> the uses of an aggregate Vector2 is not particularly helpful

I second this. It's cool to use aggregates when possible but we have to jump through too many hoops to make that happen for Vector2 so I don't think we'll make such a change. We can ignore that as a potential SFML 4 change.

**Comment by TerensTare** [comment]
> It's cool to use aggregates when possible but we have to jump through too many hoops to make that happen for Vector2 so I don't think we'll make such a change.

I understand the decision and have nothing against it. I was simply trying to show what can be done to make Vector2 an aggregate if you want to, whether or not Vector2 should be an aggregate is up to you to decide.

**Comment by vittorioromeo** [comment]
Here's an idea I'm experimenting with in my fork, `sf::WindowSettings`:
https://github.com/vittorioromeo/VRSFML/blob/master/include/SFML/Window/WindowSettings.hpp

Basically, it relies on C++20 designated intializers to improve the window creation API, like this:

```cpp
sf::RenderWindow window({.size{600u, 600u},
                         .title = "SFML Stencil",
                         .style = sf::Style::Titlebar | sf::Style::Close,
                         .state = sf::State::Windowed,
                         .contextSettings{.depthBits = 0, .stencilBits = 8}});
```

**Comment by ZXShady** [comment]
@vittorioromeo 

I don't see how this improves the API much as the window constructor has different arguments of different types it s pretty clear which argument is

**Comment by RoyAwesome** [comment]
> Return `std::span` instead of `const std::vector&`

I'd be careful about this, because you can create a span to the underlying memory in a vector, and if that memory is ever reallocated or moved, you will have a span viewing moved memory.  Storing a `std::span` you dont own the underlying memory of doesn't feel as weird/dangerous as storing a `const std::vector&` that you don't own, and users would find bugs/issues doing this.  The problem is, of course, on the user (dont store spans!), but smart api design can mitigate these issues.  

Passing in `std::span` instead of `const std::vector&` is fine tho as long as you dont store the std::span.

**Comment by ChrisThrasher** [comment]
> I'd be careful about this, because you can create a span to the underlying memory in a vector, and if that memory is ever reallocated or moved, you will have a span viewing moved memory

Thanks, that's good context to keep in mind.

**Comment by ChrisThrasher** [comment]
> Hi there, what about shipping the library as c++20 modules? Don't know if it is feasible or worth but i am excited looking for that

It's not feasible given the state of C++20 modules tooling. Modules are usable for applications (which implies you're not installing any libraries) but there are still unsolved problems with regard to shipping libraries that are modularized.

**Comment by RougherO** [comment]
> Replace `sf::U8string` with `std::u8string`

I would like to point out that C++20 brings really flawed incomplete support for `std::u8string`, `std::string` seems to be able to handle unicode without any problem, although I am basing this fact off of Linux, for Windows what I seem to gather from [SO](https://stackoverflow.com/questions/56833000/c20-with-u8-char8-t-and-stdstring) is that MSVC causes problems.
[source](https://godbolt.org/z/Ph5jMxMaj)

**Comment by RougherO** [comment]
Also another possible addition could be using `std::variant` for different shapes and using `std::visit` dispatch instead of virtual dispatch for the draw calls, this becomes very easy using something like deducing this to achieve compile time polymorphism

**Comment by kimci86** [comment]
@RougherO Users are allowed to derive `sf::Shape` class on purpose. Using `std::variant` for different shapes would break that.

**Comment by RougherO** [comment]
> @RougherO Users are allowed to derive `sf::Shape` class on purpose. Using `std::variant` for different shapes would break that.

That's true, in that case how about converting sf::Shape(I think sf::Drawable will be more suitable here) to use compile time polymorphism instead of virtual functions, something like CRTP? Is there any reason for why we have virtual functions?

It would take minimal change in C++23:
```cpp
// Example
template <typename Derived>
void draw(this Derived&& self, sf::RenderTarget& target, sf::RenderState state) const
{
    std::forward<Derived>(self).draw(target, state);
}
```

**Comment by ChrisThrasher** [comment]
> Is there any reason for why we have virtual functions?

Because virtual functions get the job done and I'm not aware of a better solution that is C++17-compliant.

**Comment by eXpl0it3r** [comment]
We can also start using `explicit(false)` to indicate an intentional implicit conversion

**Comment by raustin9** [comment]
Not sure how helpful this is, but you can also do
```cpp
class [[nodiscard]] MyClass
{
};
```
to mark all methods as [[nodiscard]] for a type

**Comment by ChrisThrasher** [comment]
Marking a class as `[[nodiscard]]` is a different thing. It does not implicitly mark all methods as `[[nodiscard]]`. Besides, that’s still a C++ 17 so it’s not relevant to this thread.

**Comment by raustin9** [comment]
Mixing up my standards I see. And yeah I guess I had a misconception about that. C++20 does allow you to do this though:
```cpp
[[nodiscard("Returns whether the action failed")]]
Error performAction() {...}
```
which can be helpful especially for functions on the API boundary
